content/blog/overview-package-management-python.md

```diff
@@ -0,0 +1,363 @@
+not able to give insights on that side of the package management story.
+
+I will begin by quickly explaining the most common way of setting up Python projects and handling dependencies
+to make sure everyone is on the same page.
+
+## A basic setup
+
+Pretty much all the Django/Flask projects I have seen use the same structure: a `requirements`
+folder containing several `.txt` files listing the dependencies and their versions.
+
+The folder typically looks like the following:
+
+```bash
+.
+├── base.txt
+├── local.txt
+└── prod.txt
+```
+
+`base.txt` contains all the essential dependencies:
+
+```txt
+boto3==1.6.10
+
+Django==2.0.3
+django-cors-headers==2.2.0
+django-filter==1.1.0
+django-floppyforms==1.7.0
+django-nested-inline==0.3.7
+django-storages==1.6.5
+djangorestframework==3.7.7
+djangorestframework-jsonp==1.0.2
+social-auth-app-django==2.1.0
+social-auth-core==1.7.0
+django-widget-tweaks==1.4.1
+```
+
+While it is not required, every requirement files I have seen were using `==` to
+depend on the exact versions to ensure reproducibility. Well, as much reproducibility
+as a mutable package index allows anyway as we will see later.
+
+Other requirements files will import the `base.txt` file and then add environment specific dependencies such
+as testing libraries or developer tools for `local.txt` for example.
+
+```txt
+-r base.txt
+
+ipython==6.2.1
+flake8==3.5.0
+django-extensions==2.0.6
+mypy==0.570
+factory_boy==2.10.0
+```
+
+But wait! Before being able to install the dependencies, you first need to create
+a [virtual environment](https://docs.python.org/3/tutorial/venv.html). Python 3
+has built-in support for virtualenvs and you can create one simply by running `python3 -m venv tutorial-env`.
+To activate it, you need to execute the `activate` script in the `bin` folder of the virtualenv.
+
+As this feature is somewhat recent, many virtualenv helpers exist to make it easier to use.
+I use [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/) combined with the `zsh` plugin with
+the same name to automatically activate the right virtualenv when I'm `cd`ing into a folder that has one.
+
+Now that your virtualenv is setup and activated, you can run `pip install -r requirements/local.txt`
+to install your dependencies.
+
+You can list the currently installed packages by running `pip freeze` in the virtualenv. The downside is that it will also show
+all indirect dependencies as well. You can order that list with the `-r my_file.txt` argument to bubble your explicit dependencies
+at the top but it isn't great.
+
+This process works but is quite manual:
+
+- beginners might not know of virtualenvs and how to use them
+- packages and their versions are edited manually so it is easy to make mistakes
+by forgetting to add/remove one
+- not really secure: the package index [Pypi](https://pypi.org) is mutable so 2.1.0 one day
+can be different from 2.1.0 another day and there will no way to know about it. While you can add
```

Comment:
"no way to know about it": There's plenty ways to know about it.
We used to run regularly a command like:
```
$ docker-compose exec backend pip list -o --format=columns|grep -E "$(cat path/to/project/requirements*|awk '{split($0, a, /(>|=)/); print a[1]}'|tr '\n' '|')echo ${query%?}"
```

Then we adopted a github tool called [renovate](https://renovatebot.com/) and since it auto-creates PR for us, all we have is reading the release notes printed in the PR description then accept and merge.

Reply:
Oh sorry, after re-reading I get what you mean. Well... rebuilding packages on the same version is a very bad practice. It's up to us to verify and request that the packages we use follow a proper semantic (have been using [semver](https://semver.org/) for years now)

Reply:
sure, but i can delete django 2.1.1 and reupload a new version. Without hashes, you will not realise that.

> It's up to us to verify and request that the packages we use follow a proper semantic (have been using semver for years now)
Django doesn't follow semantic versioning and many other libs don't do it either (pytz etc) but that's not the issue anyway.